spec/OpenLineage.md

```diff
@@ -0,0 +1,110 @@
+
+- **Run**: An instance of a running job with a start and completion (or failure) time. It is uniquely identified by an id relative to its job definition.
+
+- **Facet**: A piece of metadata attached to one of the entities defined above.
+
+example:
+Here is an example of a simple start run event not adding any facet information:
+```
+{
+  "transition": "START",
+  "eventTime": "2020-12-09T23:37:31.081Z",
+  "run": {
+    "runId": "345",
+  },
+  "job": {
+    "namespace": "my-scheduler-namespace",
+    "name": "myjob.mytask",
+  },
+  "inputs": [
+    {
+      "namespace": "my-datasource-namespace",
+      "name": "instance.schema.table",
+    }
+  ],
+  "outputs": [
+    {
+      "namespace": "my-datasource-namespace",
+      "name": "instance.schema.output_table",
+    }
+  ],
+  "producer": "https://github.com/OpenLineage/OpenLineage/blob/v1-0-0/client"
+}
+```
+
+### Lifecycle
+
+The OpenLineage API defines events to capture the lifecycle of a *Run* for a given *Job*.
+When a *job* is being *run*, we capture metadata by sending run events when the state of the job transitions to a different state.
+We might observe different aspects of the job run at different stages. This means that different metadata might be collected in each event during the lyfecycle of a run.
+All metadata is additive. for example, if more inputs or outputs are detected as the job is running we might send additional events specifically for those datasets without re-emiting previously observed inputs or outputs.
+Example:
+ - When the run starts, we collect the following Metadata:
+    - Run Id
+    - Job id
+    - transition: START
+    - event time
+    - source location and version (ex: git sha)
+    - If known: Job inputs and outputs. (input schema, ...)
+ - When the run completes:
+    - Run Id
+    - Job id
+    - transition: COMPLETE
+    - event time
+    - Output datasets schema (and other metadata).
+
+
+
+### Facets
+
+Facets are pieces of metadata that can be attached to the core entities:
+- Run
+- Job
+- Dataset
+
+A facet is an atomic piece of metadata identified by its name. This means that emiting a new facet whith the same name for the same entity replaces the previous facet instance for that entity entirely). It is defined as a JSON object that can be either part of the spec or custom facets defined in a different project. 
+
+Custom facets must use a distinct prefix named after the project defining them to avoid colision with standard facets defined in the [OpenLineage.yml](OpenLineage.yml) OpenAPI spec. 
+They have a schemaURL field pointing to the corresponding version of the facet schema (as a [$ref URL location](https://swagger.io/docs/specification/using-ref/) ).
+
+Example: https://github.com/OpenLineage/OpenLineage/blob/v1-0-0/spec/OpenLineage.yml#MyCustomJobFacet
+
+Custom facets can be promoted to the standard by including them in the spec.
```

Comment:
What isn't very clear to me is how facet information is added by providers. For example, when Iceberg commits to a table, we get lots of information like what partitions were changed, what the operation was (append, overwrite, etc.), how long the commit took, number of retries, etc. That information isn't exposed to Spark, which has a simple API that is generic across sources. How would Iceberg attach the facet information?

I can think of a couple options:
1. Add facets to Spark's source API
2. Send more lineage events directly from Iceberg
3. Add calls in Iceberg to pass facet information to some listener

I don't think that option 1 is a good idea because it would add complication to the API and would need to be done in every compute engine.

Option 2 is also not a great option because it isn't clear exactly what the "run" is in this case. I think that a run would probably be at the level of a SQL statement here, and further divisions seem like hacky ways to work around the problem. It isn't clear what the Iceberg commit "run" would use as its "job", for example.

Option 3 sounds more reasonable to me, but is really more a question of what the JVM API looks like.

Reply:
That's a good point, I have another PR with a code generation draft as the beginning of a java client.
https://github.com/OpenLineage/OpenLineage/pull/3
We should explore what such an API looks like.